File: 根据fasta和gff提取蛋白质序列.py
#处理fasta文件

import linecache
import re
import sys

a= linecache.getlines('GCF_000001405.39_GRCh38.p13_genomic.fna') #返回一个字符串数组
dict={} #承装染色体和序列的映射关系

theStartPosition=[] #用于记录所有染色体起始行的行号
indexToName=[] #完成theStartPosition下标（即染色体序号）到染色体名字的映射


for i in range(0,len(a)):
	if re.match(r'>',a[i]):
		theStartPosition.append(i)
		indexToName.append(re.split('>| ',a[i])[1]) #把染色体的名字添加到那个映射数组中
		


for i in range(0,len(theStartPosition)-1):
	myList=[] #暂时承装某一条染色体的所有碱基序列
	for j in range(theStartPosition[i]+1,theStartPosition[i+1]):
		a[j] = a[j].upper(); #把每一行的字符串都转换成大写
		myList += list(a[j])[0:-1]
	dict[indexToName[i]]=myList

#下面是补上最后一个染色体
myList = []
for j in range(theStartPosition[-1]+1, len(a)):
	a[j] = a[j].upper(); #把每一行的字符串都转换成大写
	myList += list(a[j])[0:-1]

dict[indexToName[-1]]=myList


___________________________________
# -*- coding: utf-8 -*-
#处理gff文件并根据上一个脚本生成的字典，进行查找。
#在第一个脚本生成dict（染色体的序列字典）后，将gff文件进行处理并根据dict截取出gff，保存在一个字典cdsDict中
# import sys
import re

gff = open('GCF_000001405.39_GRCh38.p13_genomic.gff',mode='r') 

cdsDict = {} #用于装每条CDS对应的碱基序列。



def insertToDict(dict, CDS_Name, CDS_Sequence):
	if CDS_Name in dict:
		dict[CDS_Name]+= CDS_Sequence
	else:
		dict[CDS_Name] = CDS_Sequence


for line in gff:
	if re.match('#',line): #跳过#开头的行。
		continue
	myList = re.split('\t|;',line) #注意这里是用制表符匹配，如果用空格匹配的话则出错。因为文件本身就是制表符
	if myList[2]=='CDS':
		#start = int(myList[3]) + int(myList[7]) #注意文本中的数字也是字符串的，所以要先转换成int
		start = int(myList[3])
		end = int(myList[4])
		
		#if myList[8] in cdsDict:
		#	start = int(myList[3])
		#else:
		#	start = int(myList[3]) + int(myList[7]) #注意文本中的数字也是字符串的，所以要先转换成int
		#end = int(myList[4])
		
		if myList[6]=='+':
			insertToDict(cdsDict, myList[8], dict[myList[0]][start-1:end])
		else:
			if(start == 1):
				cds = dict[myList[0]][-end:]; #倒着切片然后reverse翻转(当包含最后一个元素时，就不能[-4:0]这么写了，要省略这个0，所以这里特殊处理)。
				cds.reverse()
			else:
				cds = dict[myList[0]][-end:-start+1]; #倒着切片然后reverse翻转。		
				cds.reverse()
			match = {'A':'T','T':'A','C':'G','G':'C','N':'N'}
			for i in range(0,len(cds)): #按配对方式替换
				cds[i] = match[cds[i]]
			insertToDict(cdsDict, myList[8], cds)
		

gff.close()



____________________________________________________

#直接用上一问的列表cdsList，对应成蛋白质。把蛋白质id和序列保存到字典中。

codonDict = {
    # 'M' - START, '_' - STOP
    "GCT": "A", "GCC": "A", "GCA": "A", "GCG": "A",
    "TGT": "C", "TGC": "C",
    "GAT": "D", "GAC": "D",
    "GAA": "E", "GAG": "E",
    "TTT": "F", "TTC": "F",
    "GGT": "G", "GGC": "G", "GGA": "G", "GGG": "G",
    "CAT": "H", "CAC": "H",
    "ATA": "I", "ATT": "I", "ATC": "I",
    "AAA": "K", "AAG": "K",
    "TTA": "L", "TTG": "L", "CTT": "L", "CTC": "L", "CTA": "L", "CTG": "L",
    "ATG": "M",
    "AAT": "N", "AAC": "N",
    "CCT": "P", "CCC": "P", "CCA": "P", "CCG": "P",
    "CAA": "Q", "CAG": "Q",
    "CGT": "R", "CGC": "R", "CGA": "R", "CGG": "R", "AGA": "R", "AGG": "R",
    "TCT": "S", "TCC": "S", "TCA": "S", "TCG": "S", "AGT": "S", "AGC": "S",
    "ACT": "T", "ACC": "T", "ACA": "T", "ACG": "T",
    "GTT": "V", "GTC": "V", "GTA": "V", "GTG": "V",
    "TGG": "W",
    "TAT": "Y", "TAC": "Y",
    "TAA": "_", "TAG": "_", "TGA": "_"
}

proteinSequenceDict = {} #用于装各个CDS的蛋白质序列（每个蛋白质序列是一个字符串）
#对于每一个CDS序列


for key in cdsDict:
	protein = "" #用于承装该序列的所有氨基酸
	#对于该序列中的每一个三连密码子
	for j in range(0,len(cdsDict[key]),3):
		try:
			codon = cdsDict[key][j]+cdsDict[key][j+1]+cdsDict[key][j+2] #将三联密码子相加变成字符串			
		except: #如果密码子不是3的倍数，最后数量差了的话，就break，后面这个就不算了。
			break
		try:
			protein += codonDict[codon]
		except:
			protein += '*' #  *表示未知蛋白
	proteinSequenceDict[key] = protein

File: 根据fasta和gff提取蛋白质序列.pl
#!/usr/bin/perl

#打开fna文件。把对应染色体的序列放入哈希表。key是染色体名，value是碱基序列。
open(SEQ,"<GCF_000001405.39_GRCh38.p13_genomic.fna") or die "file.txt 文件无法打开，$!";
$ori=$/; 
$/=">"; #将分隔符改成>,等会就不是每次读一行了，而是每次读的内容以>分割，即每次读一个染色体

%chromNameToSeq = (); #初始化哈希表，用于存放染色体名和染色体序列的对应关系
while(<SEQ>){
	next if (/^>$/); #就是如果每次以>分隔来读入的话，那么每次读入内容的结尾就是>。由于该文件第一个字符就是>,所以第一次读入的内容就是一个>,所以需要跳过第一次。
	/(.*?)\s.*?\n(.*)/s; #读入的内容是放到变量$_中，不写对哪个匹配就默认对它匹配。该正则表达式的结尾修饰符s是指让.可以匹配换行\n，默认是不能匹配的。
	#注意正则匹配时，一个单位匹配到哪里，取决于它的下一个单位是啥以及是否贪婪。加上问号就代表不贪婪：即只要遇到它的下一个单位就停止，否则是遇到最远的下一个单位才停止
	#那么上面是先匹配任意直到遇到空格（非贪婪），那么可见匹配的就是染色体的名字；然后再匹配任意直到换行（非贪婪），就是匹配染色体名字后面的那串解释，匹配完这一整行；
	#接着是匹配任意字符（贪婪），即匹配后面所有的碱基序列（注意其中混有换行符了，等会需要去掉）
	$nameOfChromosome = $1;
	$seqOfChromsome = $2;
	$seqOfChromsome =~ s/\n//g;  #这里就是把所有的换行符\n替换成空。结尾g代表全局替换，即把所有的换行符都替换。
	$seqOfChromsome =~ tr/>//;   #这是把结尾的>删去。
	$seqOfChromsome = "\U$seqOfChromsome"; #在字符串前加上\U 即代表全转换成大写
	$chromNameToSeq{$nameOfChromosome} = $seqOfChromsome; #然后是把对应的染色体名字作为哈希表的key，对应的序列作为value（这哈希表也可以不用初始化直接用，狗得很）
}
close (SEQ);
$/=$ori; #恢复默认分隔符的值

#将标准密码子表存入哈希表condonTable1
%condonTable1=('TCA' => 'S','TCC' => 'S','TCG' => 'S','TCT' => 'S','TTC' => 'F',
                'TTT' => 'F','TTA' => 'L','TTG' => 'L','TAC' => 'Y','TAT' => 'Y',
                'TAA' => '*','TAG' => '*','TGC' => 'C','TGT' => 'C','TGA' => '*',
                'TGG' => 'W','CTA' => 'L','CTC' => 'L','CTG' => 'L','CTT' => 'L',
                'CCA' => 'P','CCC' => 'P','CCG' => 'P','CCT' => 'P','CAC' => 'H',
                'CAT' => 'H','CAA' => 'Q','CAG' => 'Q','CGA' => 'R','CGC' => 'R',
                'CGG' => 'R','CGT' => 'R','ATA' => 'I','ATC' => 'I','ATT' => 'I',
                'ATG' => 'M','ACA' => 'T','ACC' => 'T','ACG' => 'T','ACT' => 'T',
                'AAC' => 'N','AAT' => 'N','AAA' => 'K','AAG' => 'K','AGC' => 'S',
                'AGT' => 'S','AGA' => 'R','AGG' => 'R','GTA' => 'V','GTC' => 'V',
                'GTG' => 'V','GTT' => 'V','GCA' => 'A','GCC' => 'A','GCG' => 'A',
                'GCT' => 'A','GAC' => 'D','GAT' => 'D','GAA' => 'E','GAG' => 'E',
                'GGA' => 'G','GGC' => 'G','GGG' => 'G','GGT' => 'G');

#将第二套密码子表存入哈希表condonTable2
%condonTable2=('TCA' => 'S','TCC' => 'S','TCG' => 'S','TCT' => 'S','TTC' => 'F',
                'TTT' => 'F','TTA' => 'L','TTG' => 'L','TAC' => 'Y','TAT' => 'Y',
                'TAA' => '*','TAG' => '*','TGC' => 'C','TGT' => 'C','TGA' => 'W',#由*变为W
                'TGG' => 'W','CTA' => 'L','CTC' => 'L','CTG' => 'L','CTT' => 'L',
                'CCA' => 'P','CCC' => 'P','CCG' => 'P','CCT' => 'P','CAC' => 'H',
                'CAT' => 'H','CAA' => 'Q','CAG' => 'Q','CGA' => 'R','CGC' => 'R',
                'CGG' => 'R','CGT' => 'R','ATA' => 'M',#由I变为M
                'ATC' => 'I','ATT' => 'I',
                'ATG' => 'M','ACA' => 'T','ACC' => 'T','ACG' => 'T','ACT' => 'T',
                'AAC' => 'N','AAT' => 'N','AAA' => 'K','AAG' => 'K','AGC' => 'S',
                'AGT' => 'S','AGA' => '*',#由R变为*
                'AGG' => '*',#由R变为*
                'GTA' => 'V','GTC' => 'V',
                'GTG' => 'V','GTT' => 'V','GCA' => 'A','GCC' => 'A','GCG' => 'A',
                'GCT' => 'A','GAC' => 'D','GAT' => 'D','GAA' => 'E','GAG' => 'E',
                'GGA' => 'G','GGC' => 'G','GGG' => 'G','GGT' => 'G');

#打开gff文件。匹配到CDS的行，然后根据起止位置、正负链、CDS的id（相同id的CDS拼到一起），从刚才那个染色体序列哈希表中去查找对应的序列，然后把CDS名和序列存入哈希表。
open(ANNO,"<GCF_000001405.39_GRCh38.p13_genomic.gff") or die "file.txt 文件无法打开，$!";
%useWhichTable=(); #初始化一个哈希表，用于记录某个CDS名是用哪一个密码子表
%cdsNameToSeq=();  #初始化一个哈希表，用于记录CDS名与碱基序列的对应关系

while(<ANNO>){
	#chomp; #每次读一行（默认按\n分隔，所以结尾是\n, 因此需要用chomp去掉。省略参数就是默认对$_操作）
    if(/ID=cds-(.*?);/){  #如果匹配到cds的话，说明这一行是cds。小括号里就匹配-之后的任意个字符直到分号，（非贪婪匹配），因此匹配的就是cds名。
        my @line=split("\t",$_); #将匹配到的这一行用制表符分割成数组
        $useWhichTable{$1}=index($_,"transl_table=2"); #$1就是匹配到的CDS名，看它对应的是哪个密码子表，并记录。
		#上面那个index是查找含有"transl_table=2"的下标，找不到返回-1就说明是第一个表，
        if($line[6] eq "-"){
			$aPartOfSeq=reverse(substr($chromNameToSeq{$line[0]},$line[3]-1,$line[4]-$line[3]+1)); #substr就是从某个数组中，提取从第几号位置开始的n个字符。
			$aPartOfSeq=~tr/TCGA/AGCT/; #正则替换。
		}
        else{
          $aPartOfSeq=substr($chromNameToSeq{$line[0]},$line[3]-1,$line[4]-$line[3]+1);
        }
        $cdsNameToSeq{$1}.=$aPartOfSeq; #将提取到的序列拼过来。（perl中的.=就相当于字符串的+=，这也太妖了吧。。。）
    }
}
close (ANNO);

#将CDS序列存入文件cdsNameToSeq.txt
open(IDSEQ,">cdsNameToSeq.txt"); #>代表写入的意思
foreach $key (keys %cdsNameToSeq){    #keys 加哈希表 再用括号括起来，外面foreach，这种语法到时候再记一下。太weird。
	print IDSEQ ">$key\n$cdsNameToSeq{$key}\n";  #print + 文件句柄 + 内容 就是把内容输入到文件。（这种写法狗得很，让我一个java过来的很是不爽）
}
close(IDSEQ);

#将CDS序列翻译为蛋白质序列存入哈希%cdsToProtenSeq
%cdsToProtenSeq=();#初始化哈希表，存放CDS名对应的蛋白质序列

foreach $key (keys %cdsNameToSeq){ #对于每个CDS序列，在之后再弄一层循环，一个个把碱基翻译成蛋白质，然后拼在一起
    $protein=""; #初始蛋白质是空，之后一个个拼过来
    if($useWhichTable{$key}==-1){ #如果用的是第一个密码子表
		for(my $a=0;$a<length($cdsNameToSeq{$key})-2;$a+=3){    #沿着碱基序列三个三个往后走
        $protein.=$condonTable1{substr($cdsNameToSeq{$key},$a,3)};  #substr就是对于某一段字符串，把它第n个位置的3个碱基提出来
        }
    }
    else{
		for(my $a=0;$a<length($cdsNameToSeq{$key})-2;$a+=3){
        $protein.=$condonTable2{substr($cdsNameToSeq{$key},$a,3)};
        }
    }
    $cdsToProtenSeq{$key}=$protein; 
    print  ">$key\n$cdsToProtenSeq{$key}\n";  #把CDS名和对应的蛋白质序列输出看一下。注意print后跟双引号，默认结尾没有换行。所以需要手动加上一个换行。
}

#将蛋白序列存入文件cdsToProtenSeq.txt
open(IDPRO,">cdsToProtenSeq.txt");
foreach $key (keys %cdsToProtenSeq){
    print IDPRO ">$key\n$cdsToProtenSeq{$key}\n";
}
close IDPRO;
